aspnetcore/blazor/security/webassembly/standalone-with-identity/qrcodes-for-authenticator-apps.md

```diff
@@ -0,0 +1,853 @@
+                                <InputText type="password" 
+                                    @bind-Value="Input.Password" 
+                                    id="Input.Password" 
+                                    class="form-control" 
+                                    autocomplete="current-password" 
+                                    aria-required="true" 
+                                    placeholder="password" />
+                                <label for="Input.Password" class="form-label">
+                                    Password
+                                </label>
+                                <ValidationMessage For="() => Input.Password" 
+                                    class="text-danger" />
+                            </div>
+                        </div>
+                        <div style="display:@(requiresTwoFactor ? "block" : "none")">
+                            <div class="form-floating mb-3">
+                                <InputText @bind-Value="Input.TwoFactorCode" 
+                                    id="Input.TwoFactorCode" 
+                                    class="form-control" 
+                                    autocomplete="off" 
+                                    placeholder="###### or #####-#####" />
+                                <label for="Input.TwoFactorCode" class="form-label">
+                                    Two-factor Code or Recovery Code
+                                </label>
+                                <ValidationMessage For="() => Input.TwoFactorCode" 
+                                    class="text-danger" />
+                            </div>
+                        </div>
+                        <div>
+                            <button type="submit" class="w-100 btn btn-lg btn-primary">
+                                Log in
+                            </button>
+                        </div>
+                        <div class="mt-3">
+                            <p>
+                                <a href="forgot-password">Forgot password</a>
+                            </p>
+                            <p>
+                                <a href="register">Register as a new user</a>
+                            </p>
+                        </div>
+                    </EditForm>
+                </section>
+            </div>
+        </div>
+    </NotAuthorized>
+</AuthorizeView>
+
+@code {
+    private FormResult formResult = new();
+    private bool requiresTwoFactor;
+    private string? recoveryCodesRemainingMessage;
+
+    [SupplyParameterFromForm]
+    private InputModel Input { get; set; } = new();
+
+    [SupplyParameterFromQuery]
+    private string? ReturnUrl { get; set; }
+
+    public async Task LoginUser()
+    {
+        if (requiresTwoFactor)
+        {
+            if (!string.IsNullOrEmpty(Input.TwoFactorCode))
+            {
+                if (Input.TwoFactorCode.Length == 6)
```

Comment:
```suggestion
                // The [RegularExpression] data annotation ensures that the input will be a six-digit
                // authenticator code (######) or eleven-character alphanumeric recovery code (#####-#####)
                if (Input.TwoFactorCode.Length == 6)
```

My immediate reaction when seeing this code is that it would too easily interpret a fat fingered 2FA code as a recovery code. That was until I saw the regex, so I think it's worth calling out in a comment why just comparing the length to 6 is okay.